File: pidex/core/dedup.py
from pidex.core.event import Event


class DedupManager:
    def __init__(self):
        self._last_keys: dict[str, str] = {}

    def is_new(self, event: Event) -> bool:
        key = event.dedup_key()
        last = self._last_keys.get(event.source)
        return last != key

    def record(self, event: Event) -> None:
        self._last_keys[event.source] = event.dedup_key()

    def reset(self, source: str) -> None:
        self._last_keys.pop(source, None)

File: internal/core/dedup.go
package core

import "sync"

type DedupManager struct {
	mu       sync.Mutex
	lastKeys map[string]string
}

func NewDedupManager() *DedupManager {
	return &DedupManager{lastKeys: make(map[string]string)}
}

func (m *DedupManager) IsNew(event Event) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	key := event.DedupKey()
	last, ok := m.lastKeys[event.Source]
	if !ok {
		return true
	}
	return last != key
}

func (m *DedupManager) Record(event Event) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.lastKeys[event.Source] = event.DedupKey()
}

func (m *DedupManager) Reset(source string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.lastKeys, source)
}
